scanner/src/main/kotlin/scanners/fossid/FossIdUrlProvider.kt

```diff
@@ -31,40 +32,109 @@ import org.ossreviewtoolkit.utils.ort.requestPasswordAuthentication
  * An internal helper class that generates the URLs used by [FossId] to check out the repositories to be scanned.
  *
  * The URLs used by FossId can sometimes be different from the normal package URLs. For instance, credentials may need
- * to be added. This class takes care of such mappings.
+ * to be added, or a different protocol may be used. This class takes care of such mappings.
  */
 internal class FossIdUrlProvider private constructor(
-    /** Flag whether credentials should be passed to FossID in URLs. */
-    val addAuthenticationToUrl: Boolean
+    /**
+     * The URL mapping. URLs matched by a key [Regex] are replaced by the URL in the value. The replacement string can
+     * refer to matched groups of the [Regex]. It can also contain the variables [VAR_USERNAME] and [VAR_PASSWORD] to
+     * allow the injection of credentials.
+     */
+    private val urlMapping: Map<Regex, String>
 ) {
     companion object : Logging {
-        fun create(addAuthenticationToUrl: Boolean): FossIdUrlProvider = FossIdUrlProvider(addAuthenticationToUrl)
+        /** Variable in the mapping replacement string that references the username. */
+        private const val VAR_USERNAME = "#user"
```

Comment:
I guess you deliberately chose `#` over `$` as a marker here, but if `$` works I guess I would prefer that, so the user does not have to remember that other marker char, and because "#user" might actually be an anchor a URL (very unlikely for a Git repo URL, though).

Reply:
Yes, I wanted to avoid conflicts with the named groups in the regular expression. And indeed: when changing the variables to use the `$` prefix, the replace operation fails with:
`java.lang.IllegalArgumentException: Illegal group reference`

Reply:
Ok, thanks for checking; then let's stick to `#`.